components/script/dom/node.rs

```diff
@@ -1166,60 +1195,75 @@ impl FollowingNodeIterator {
         self.next_skipping_children_impl(current)
     }
 
-    fn next_skipping_children_impl(&mut self, current: Root<Node>) -> Option<Root<Node>> {
+    fn next_skipping_children_impl(&mut self, current: Root<Node>) -> Option<Root<T>> {
         if self.root == current {
             self.current = None;
             return None;
         }
 
-        if let Some(next_sibling) = current.GetNextSibling() {
-            self.current = Some(next_sibling);
-            return current.GetNextSibling()
+        while let Some(next_sibling) = current.GetNextSibling() {
+            if let Some(next_sibling) = Root::downcast::<T>(next_sibling) {
+                self.current = Some(Root::upcast::<Node>(next_sibling.clone()));
+                return Some(next_sibling.clone())
+            }
         }
 
         for ancestor in current.inclusive_ancestors() {
             if self.root == ancestor {
                 break;
             }
-            if let Some(next_sibling) = ancestor.GetNextSibling() {
-                self.current = Some(next_sibling);
-                return ancestor.GetNextSibling()
+            while let Some(next_sibling) = ancestor.GetNextSibling() {
+                if let Some(next_sibling) = Root::downcast::<T>(next_sibling) {
+                    self.current = Some(Root::upcast::<Node>(next_sibling.clone()));
+                    return Some(next_sibling.clone())
+                }
             }
         }
         self.current = None;
         None
     }
 }
 
-impl Iterator for FollowingNodeIterator {
-    type Item = Root<Node>;
+impl<T> Iterator for FollowingIterator<T> where T: DerivedFrom<Node> {
+    type Item = Root<T>;
 
     // https://dom.spec.whatwg.org/#concept-tree-following
-    fn next(&mut self) -> Option<Root<Node>> {
+    fn next(&mut self) -> Option<Root<T>> {
         let current = match self.current.take() {
             None => return None,
             Some(current) => current,
         };
 
-        if let Some(first_child) = current.GetFirstChild() {
-            self.current = Some(first_child);
-            return current.GetFirstChild()
+        while let Some(first_child) = current.GetFirstChild() {
+            if let Some(first_child) = Root::downcast::<T>(first_child) {
+                self.current = Some(Root::upcast::<Node>(first_child.clone()));
+                return Some(first_child);
+            }
+        }
+
+        while let Some(first_child) = current.GetFirstChild() {
+            if let Some(first_child) = Root::downcast::<T>(first_child) {
+                self.current = Some(Root::upcast::<Node>(first_child.clone()));
+                return Some(first_child);
+            }
```

Comment:
I don't understand what you are trying to do by repeating the loop, and `current` is still never yielded again.